### ISSUE TITLE
Case and dictionary/parameter keys

### ISSUE THREAD

**Issue by mnot** [httpwg/http-extensions#1209]
One of the bits of feedback we got in IESG review was regarding the case sensitivity of dictionary and parameter keys. I think we dealt with the immediate issue, but it reminded me of a concern that's been in the back of my mind for a while -- that the forced-lowercase nature of these keys might be surprising and unwelcome, especially for developers who are encultured to use camelCase. 

That was increased by seeing the Signatures presentation at the interim this week, where the proposal had a camelCased key, even though it was supposed to be a structured field.

The most obvious place that this would have effect is if we tried to backport `Set-Cookie` into a Structured Field, because it has a number of camelCase parameters. I had a chat with @mikewest who reassuringly pointed out that `Set-Cookie` parameters are case-insensitive, and they have tests to prove it. 

Still, I have a lingering concern here, and I wanted to clear it up before we publish the spec. I think the options are:

1. Make editorial adjustments to make it super-clear that keys are case-forced to lowercase.

2. Open up keys to allow uppercase characters, both on the wire and in the data model.

3. Change the parsing algorithm to accept uppercase characters, but to force them to lowercase in the data model.

If we do (1), we can update the spec and ship it. The other options would require taking it to the mailing list, getting consensus and then updating the spec, tests and implementations.

Thoughts? I'm personally hovering between (1) and (3); I don't think (2) is good, in that it will make keys case-sensitive. I think the question, then, is how much appetite we have for change at this late stage.

/cc @clelland @phluid61 @bsdphk

**Comment by reschke** [comment]
I don't see how 3) is helpful, unless you have parsing existing fields in mind.

1) Works for me, so would 2).

**Comment by LPardue** [comment]
FWIW, I recently reviewed the CTA-5004 draft for Common-Media-Client-Data which includes a request field definition. In the context of case sensitivity, it appears that document has competing statements of definition that I cherry picked below; note "payload" effectively means field value in this case

> All information in the payload MUST be represented as <key>=<value> pairs.

> All key names are case-sensitive.

> The data payload syntax is intended to be compliant with Structured Field Values for HTTP [7].

I initially raised an issue about sensitivity https://github.com/cta-wave/common-media-client-data/issues/37 without appreciating the problem with Structured Field

**Comment by bsdphk** [comment]
I'm firmly for option one.

**Comment by mnot** [comment]
(3) would be similar to how field names work in HTTP/2; it allows people to use camelCase etc. if they wish to informally, but in the data model, it's forced to be case insensitive.

I don't know how successful (1) will be. That's why my preference is (3), although I can live with (1) if there's strong resistance.

Anyone else with an opinion? This is the only thing preventing us from shipping.

**Comment by wtarreau** [comment]
Well, IMHO (3) doesn't provide much value over (1) for the long term, but still requires case-insensitive processing that's more painful and error-prone.

**Comment by kazuho** [comment]
+1 to option (1).

There is no benefit in allowing new specifications using Structured Headers to use CamelCase. 

When converting existing header fields to Structured Headers, assuming that those existing header fields use case-insensitive names, we can require the transformation process from existing format to the Structured Headers to lower-case the names. Requiring such conversion is no more complex than dealing with case-sensitivity on the receiver side.

To put it differently, option (3) does not help supporting existing specifications, while it adds complexity when only supporting new specifications.

Option (2) does help specifications that are CamelCased, but we've already agreed to not support them.

**Comment by phluid61** [comment]
I'm for 1.

* This is a new thing, so there's no backward compatibility issue, only old habits. </ivory-tower>
* Keys aren't strings, they're atoms. Saying that `foo` is the same as `Foo` is kind of arbitrary, and reinforces wrong-thinking. </even-taller-ivory-tower>
* I'm a bit 👎 for 3 because it's rewriting data. We rewrite invalid inputs when serialising, but the only time we break the deserialise-serialise roundtrip is to erase booleans. And it feels like erasing is less egregious than modifying.

**Comment by ioggstream** [comment]
+1 for 1. Moreover:

- if a spec switches to structured-headers,  it should probably update keys to be lowercase.

**Comment by tigertoes** [comment]
I'm in favour of option 1 as no existing draft is mandating any keys that are uppercase, but this decision might require editorial changes to draft-ietf-httpbis-variants.

Firstly, looking at all drafts that I could find that contain reference to the structured headers draft:

```bash
$ rsync -avz rsync.tools.ietf.org::tools.id ./id && cd id
$ ag --nocolor -l -G '\.txt$' 'I-D.ietf-httpbis-header-structure' \
	| sed 's=-[0-9][0-9].txt==' \
	| sort -u
```
Then looking at each draft at its latest version, asserting if either for a dictionary or a parameter which might be affected.

|                 Draft Name                 | Dictionary? | Keys?  |
|--------------------------------------------|-------------|--------|
| draft-cedik-http-warning                   | ❌ |       N/A       |
| draft-gruessing-sdp-http                   | ⭕️ | Lower-case only |
| draft-ietf-httpbis-bcp56bis                | ❌ |       N/A       |
| draft-ietf-httpbis-cache-header            | ⭕️ | Lower-case only |
| draft-ietf-httpbis-client-hints            | ❌ |       N/A       |
| draft-ietf-httpbis-proxy-status            | ⭕️ | Lower-case only |
| draft-ietf-httpbis-variants                | ⭕️ | See below       |
| draft-kazuho-httpbis-priority              | ⭕️ | Lower-case only |
| draft-nottingham-bikeshed-length           | ❌ |       N/A       |
| draft-nottingham-binary-structured-headers | ❌ |       N/A       |
| draft-nottingham-cache-header              | ⭕️ | Lower-case only |
| draft-nottingham-proxy-status              | ⭕️ | Lower-case only |
| draft-ohanlon-transport-info-header        | ⭕️ | Lower-case only |
| draft-pot-prefer-push                      | ❌ |       N/A       |
| draft-thomson-http-mice                    | ❌ |       N/A       |
| draft-thomson-protocol-freedom             | ❌ |       N/A       |
| draft-west-http-state-tokens               | ⭕️ | Lower-case only |
| draft-west-lang-client-hint                | ❌ |       N/A       |
| draft-west-ua-client-hints                 | ❌ |       N/A       |
| draft-yasskin-http-origin-signed-responses | ⭕️ | Lower-case only |

With regards to draft-ietf-httpbis-variants, it [defines a `sh-dict`](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-variants#section-2) and gives multiple examples with mixed casing, but also goes on to say:

> Note that as always, field-name is case-insensitive.

This should be updated if consensus is made.

**Comment by mnot** [comment]
OK, seems like option 1 is where people want to go.

For the record, I think this is going to confuse / surprise people no matter what we write, but hopefully implementation consistency will drive them in the right direction.

**Comment by mnot** [comment]
See commit; if I don't hear any pushback soon, we'll ship this.

**Comment by clelland** [comment]
I would have had a slight preference for 3), if only for consistency with field names, but I can live with 1).

`Set-Cookie` might just have to be a special case; if we do redefine it as a structured field, then at that point the old syntax just gets deprecated, but like any other deprecated feature, clients will just continue to recognize it until usage drops below some threshold.

Or maybe it's just one of the legacy non-structured fields. Reinterpreting old fields as structured is nice-to-have, but the goal isn't to get all of them.
